GVRf/Framework/framework/src/main/jni/engine/renderer/renderer.cpp

```diff
@@ -121,7 +130,212 @@ void Renderer::state_sort() {
+                if (existing_batch && existing_batch->isBatchDirty()) {
+                    existing_batch->setMeshesDirty();
+
+                    std::unordered_map<Batch*, int>::iterator it =
+                            batch_map.find(existing_batch);
+                    if (it != batch_map.end()) {
+                        int index = it->second;
+                        batch_set.erase(batch_set.begin() + index);
+                    }
+                    delete existing_batch;
+
+                    Batch* new_batch = new Batch(MAX_INDICES, MAX_INDICES);
+                    new_batch->add(render_data);
+                    render_data->setBatch(new_batch);
+                    batch_set.push_back(new_batch);
+                    batch_map[new_batch] = batch_set.size() - 1;
+                    existing_batch = new_batch;
+                }
+                // existing batch is not full
+                else if (existing_batch
+                        && existing_batch->getNumberOfMeshes() < BATCH_SIZE) {
+
+                    if (!existing_batch->add(render_data)) {
+                        Batch* new_batch = new Batch(MAX_INDICES, MAX_INDICES);
+                        new_batch->add(render_data);
+                        render_data->setBatch(new_batch);
+                        batch_set.push_back(new_batch);
+                        batch_map[new_batch] = batch_set.size() - 1;
+                        existing_batch = new_batch;
+                    }
+
+                    render_data->setBatch(existing_batch);
+
+                    if (batch_map.find(existing_batch) == batch_map.end()) {
+                        batch_set.push_back(existing_batch);
+                        batch_map[existing_batch] = batch_set.size() - 1;
+                    }
+                } else { //existing batch is full or not exists
+                    Batch* new_batch = new Batch(MAX_INDICES, MAX_INDICES);
+                    new_batch->add(render_data);
+                    render_data->setBatch(new_batch);
+                    batch_set.push_back(new_batch);
+                    batch_map[new_batch] = batch_set.size() - 1;
+                    existing_batch = new_batch;
+                }
+            }
+        } else { // batch is not null
+
+            // update the transform if model matrix is changed
+            if (render_data->owner_object()->isTransformDirty()
+                    && render_data->owner_object()->transform()) {
+                current_batch->UpdateModelMatrix(render_data,
+                        render_data->owner_object()->transform()->getModelMatrix());
+            }
+
+            if (!render_data->mesh()->isDynamic()) {
+                if (batch_map.find(current_batch) == batch_map.end()) {
+                    batch_set.push_back(current_batch);
+                    batch_map[current_batch] = batch_set.size() - 1;
+                }
+            } else { // mesh is dynamic
```

Comment:
Is the code for dynamic mesh the same as in the previous if block? If it meant to be same refactor it into a common function.